e2e: don't assume NodeIP are reachable

aojea opened the pull request:
it seems that the CI somehow managed to provide connectivity between the worker running the tests and the Nodes on the cluster under test.

However, this is not always true, causing these test to fail.

For NodePort tests, we can test the functionality within the cluster, no need to test NodePorts from the e2e.test binary.


/kind failing-test
```release-note
NONE
```

I've already implemented the NodePort connectivity test from within the cluster in https://github.com/kubernetes/kubernetes/pull/114171, so we are not removing coverage here

Fixes: #114577

/assign @wojtek-t

wojtek-t commented on test/e2e/network/loadbalancer.go:
```diff
@@ -188,9 +188,6 @@ var _ = common.SIGDescribe("LoadBalancers", func() {
 
-		ginkgo.By("hitting the TCP service's NodePort")
-		e2eservice.TestReachableHTTP(ctx, nodeIP, tcpNodePort, e2eservice.KubeProxyLagTimeout)
-
```
> it seems that the CI somehow managed to provide connectivity between the worker running the tests and the Nodes on the cluster under test.
>
> However, this is not always true, causing these test to fail.
>
> For NodePort tests, we can test the functionality within the cluster, no need to test NodePorts from the e2e.test binary.


OK. So the reason is that majority of tests are running tests using public clusters, where nodes have public IPs.
Our large scale scalability tests are using private clusters, where nodes have only private IPs, which obviously means that these won't be reachable.

Back then we implemented the functionality of "bastion" (there is ~1 node that has public IP via which we can ssh to nodes).
I'm not sure if we can somehow reuse this concept here, but just pointing you to it, to double check if we can't reuse it.

wojtek-t replied:
Forgot to add a link:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/ssh/ssh.go#L53

aojea replied:
We try to keep the e2e test as much hermetic as possible, in this case NodePorts can be tested within the cluster so I prefer to avoid to test them also from an external place, that has the dependencies on a lot of external factors like Firewalls, LoadBalancer, ....

For LoadBalancers the test is realistic, since you expect your application to be reachable externally

wojtek-t replied:
>We try to keep the e2e test as much hermetic as possible,

+1

Can you point me to tests that already exercise this functionality? [Just to double check we are not missing anything subtle here...]

aojea replied:
yeah, NodePort are Conformance and they have also another tests like these that exercise them under the hood

https://testgrid.k8s.io/sig-network-kind#sig-network-kind,%20master&include-filter-by-regex=NodePort

aojea committed:
e2e NodePorts should be tested within the cluster

The NodePort functionality can be tested within the cluster.

Testing from outside the cluster assumes that there is connectivity
between the e2e.test binary and the cluster under test, that is not
always true, and in some cases is exposed to external factors  or
misconfigurations like wrong routes or firewall rules that impact
on the test.

Change-Id: Ie2fc8929723e80273c0933dbaeb6a42729c819d0

aojea commented:
/test pull-kubernetes-e2e-gce-correctness

aojea commented:
> /test pull-kubernetes-e2e-gce-correctness

this will provide signal https://testgrid.k8s.io/presubmits-kubernetes-scalability#pull-kubernetes-e2e-gce-correctness

aojea commented:
/test pull-kubernetes-e2e-kind

unrelated failure

wojtek-t commented:
/lgtm
/approve

Thanks!

aojea commented:
/hold

to wait for pull-kubernetes-e2e-gce-correctness

aojea commented:
/hold cancel

it is passing now
